### DISCUSSION TITLE
Using your own build of Alpine

### DISCUSSION THREAD

**Discussion by mokhosh** [alpinejs/alpine#2380] [1. Help] [answered]
After finding the need to build my own version of Alpine temporarily here https://github.com/alpinejs/alpine/discussions/2329 here's what I did:
- forked alpinejs repo
- cloned my forked on my local
- made the changes
- npm install
- npm run test all passing
- npm run build
- tried importing all four generated js files in the dist folder

Only the esm one compiled properly, and when I open my application I get warnings and errors and seems like many things are not defined (catch is not a function, my x-data variables aren't defined, etc.)

Am I missing a step or something?

Node v17.1.0
NPM v8.1.2

**Comment by danddanddand** [comment, marked as answer]
- The `catch is not a function` points to your transpile target needing to be higher es2017+
- In your own project, if you are using cdn version, you should just be able pull out of the dist folder. If you are bundling (webpack/parcel/vite/esbuild/whatever), you should copy the alpinejs folder from the alpine project somewhere local to project and edit your package.json to point to the local version. Something like this `"alpinejs": "file:./alpineDebug/alpinejs",`. Nuke your node_modules, and try build.

**Reply by mokhosh** [reply]
Thanks, I red that somewhere but I didn't find anything about target in mix. I'm using laravel mix, can i target es17+ with mix?

**Reply by danddanddand** [reply]
I have never used laravel mix so no clue. Looks like it just wraps webpack so you might be able to pass in target. https://github.com/laravel-mix/laravel-mix/blob/master/docs/quick-webpack-configuration.md

I've been using esbuild-loader for faster builds with webpack, so I have 

```
{
                    exclude: /node_modules/,
                    test: /\.ts(x?)$/,
                    loader: 'esbuild-loader',
                    options: {
                        loader: 'tsx',
                        target: 'es2016'
                    },
                }
}
```

in our project. Important part is the target being high enough. es2016 or higher. You'll probably have to see what laravel mix is using under the hood. You can npm i esbuild-loader and try that config (just switch out ts for js if you don't use typescript)

**Reply by SimoTod** [reply]
@danddanddand It's weird that es2016 works for you. The minimum target for async catch should be `es2017` are you on the last version? The requirement started from v3.5.0 I think. 

@mokhosh I don't use mix but I believe you set the target with
`mix.webpackConfig({target: 'es2017').whateverYouHaveNowInYourFile`

**Reply by danddanddand** [reply]
@SimoTod Ya, I think I started saying/thinking es2017 from you mentioning it in other discussions so I was a little surprised to see my config was using es2016. We were using the error handling months before Caleb merged it and released it in 3.5.0. It's probably something to do with having core js polyfills because of legacy devices our users were complaining about not working with our application but I should update it to higher target.

**Reply by SimoTod** [reply]
Yeah, it must be some special settings in your build process. I'm not 100% sure but I know people using phoenix had to target 2017. 🤷‍♂️

**Reply by danddanddand** [reply]
Ya, for sure strange. Our prod build used to use ts-loader and recently switched to esbuild-loader for faster webpack builds and we use Snowpack in development. We are about 90% using Alpine for components (10% are React) so the webpack stuff is a bit bloated but has been working with es2016 for both the Alpine and React code since 3.0 was released.

**Reply by mokhosh** [reply]
@SimoTod Thanks, but that didn't work. Seems like Webpack doesn't support targeting like that.

Do you recommend I use babel polyfills or should I migrate to another build tool like Rollup or Esbuild?

**Reply by danddanddand** [reply]
If it's easy to switch, I'd recommend esbuild since it's very fast and easy to use. It's also what Alpine is using. 

There for sure is a way to pass the correct configuration to Laravel Mix which is just passing through to Webpack which is just running Babel/TS/Esbuild. When you look at it that way might as well just remove two layers of abstraction and go with Esbuild directly.

**Reply by SimoTod** [reply]
weird, that's what the documentation says: https://github.com/laravel-mix/laravel-mix/blob/master/docs/api.md#webpackconfigconfig

**Reply by mokhosh** [reply]
I got this error at first:
```
[webpack-cli] Error: For the selected environment is no default script chunk format available
```
Then I fixed it like this
```
.webpackConfig({
    target: 'es2017',
    output: {
        chunkFormat: 'array-push',
    },
    resolve: {
        fallback: {
            fs: false,
        },
    },
})
```
And it's compiling, so I should be getting es2017 code, but I have the exact same error on the website.
```
Alpine Expression Error: func(...).catch is not a function
```

**Reply by mokhosh** [reply]
I asked @JeffreyWay on twitter about this issue and he said mix uses babel and babel uses browserslist, so I added this `.browserslistrc` but I get the exact same result:
```
supports async-functions
supports es6-module
supports es6
supports fetch
not IE 11
```

**Reply by mokhosh** [reply]
Here's another thing I don't understand. Why does it totally work when I import from `node_modules/alpinejs`, but crashes when I import from my own build. Shouldn't it behave the same at least?

**Reply by danddanddand** [reply]
> Here's another thing I don't understand. Why does it totally work when I import from `node_modules/alpinejs`, but crashes when I import from my own build. Shouldn't it behave the same at least?

If you start from a clean version (no node_modules) and reference your build in package.json like this, `"alpinejs": "file:./alpineDebug/alpinejs"`, the version that makes it into node_modules will be "your version" when you do npm i.

**Reply by mokhosh** [reply]
Thanks, I tried that, and it looks like a partial progress. Now I get this:
```
ERROR in ~/alpine/packages/alpinejs/dist/module.esm.js 1:0-61
Module not found: Error: Can't resolve '@babel/runtime/regenerator' in '~/alpine/packages/alpinejs/dist'
```
I searched for possible solutions, they said I should install @babel/runtime which I did even though I felt if it was necessary it should have been in Alpinejs dependencies, but that didn't help either.

**Reply by danddanddand** [reply]
you should only need the alpinejs folder, not the root of alpine

**Reply by mokhosh** [reply]
yes, that's what's being copied. just the alpinejs package.
actually one of my first tries was to copy my code manually to the node_modules folder, and it gave me the exact same error.
why wouldn't @calebporzio 's build give me the same error? should i do like a release build or something? shouldn't "npm run build" suffice?

**Reply by danddanddand** [reply]
Here is what I've done in past when I had fork I had to run

1. In fork, run npm i, make changes, run npm run build, run npm run test (usually I'd have added a test or two)
2. copy alpinejs from fork alpine/packages into my project (i used alpineDebug container folder)
3. In my project, do a clean (rm -rf node_modules, delete yarn.lock or npm lock or pnpm-lock.yaml depending what you use)
4. In my project, package.json switch `"alpinejs": "^3.5.0",` to `"alpinejs": "file:./alpineDebug/alpinejs",`
5. In my project, run npm i (or yarn or pnpm)

Basically, your forked version of Alpine should be in node_modules now and you should be able to build and use your application with the changes you made.

**Reply by mokhosh** [reply]
Ok, did all that just now.
That went back to transpiling to es5 and giving me the .catch is not a function error, and I assume it's because alpinejs is not really in node_modules and it's symlinked to alpineDebug.

**Reply by danddanddand** [reply]
It is in node_modules it's just not installing it from npm repository it's installing it from your local forked copy.

**Reply by mokhosh** [reply]
i know bro, but it doesn't copy, it just creates a symlink.
i just looked into it and it's a known issue with npm.
so the problem in both cases was creating symlink.
when i put it in a different directory because it was not in my project root folder it was not transpiled but it was also not finding necessary dependencies.
and when i put it in my project folder like you suggested webpack thought it should transpile it because it doesn't look like external code and i got the initial error.

either way, i'm stuck between extremely unfortunate decisions of npm and webpack and mix, etc.
my only option is to copy page cdn codes i think.

thanks anyway guys.

**Reply by danddanddand** [reply]
Webpack should transpile it, just not to target that mix is going via Babel. Try bundling with esbiuld instead if mix and see what it gives you.

**Reply by mokhosh** [reply]
I'm sure that will give me a working output, it's just a lot of change for one temporary fix until a new alpine release.

Thank you so much for walking me through this.
